SignalMessaging/attachments/SharingThreadPickerViewController.m

```diff
@@ -150,22 +163,142 @@ - (void)cancelShareExperience
 
 #pragma mark - AttachmentApprovalViewControllerDelegate
 
-- (void)didApproveAttachmentWithAttachment:(SignalAttachment *)attachment
+- (void)attachmentApproval:(AttachmentApprovalViewController *)attachmentApproval
+      didApproveAttachment:(SignalAttachment *)attachment
 {
     [ThreadUtil addThreadToProfileWhitelistIfEmptyContactThread:self.thread];
-    [ThreadUtil sendMessageWithAttachment:self.attachment inThread:self.thread messageSender:self.messageSender];
+    [self tryToSendAttachment:attachment fromViewController:attachmentApproval];
+}
+
+- (void)attachmentApproval:(AttachmentApprovalViewController *)attachmentApproval
+       didCancelAttachment:(SignalAttachment *)attachment
+{
+    [self cancelShareExperience];
+}
+
+#pragma mark - Helpers
 
-    // This is just a temporary hack while testing to hopefully not dismiss too early.
-    // FIXME Show progress dialog
-    // FIXME don't dismiss until sending is complete
-    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+- (void)tryToSendAttachment:(SignalAttachment *)attachment fromViewController:(UIViewController *)fromViewController
+{
+    // Reset progress in case we're retrying
+    self.progressView.progress = 0;
+
+    self.attachment = attachment;
+
+    NSString *progressTitle = NSLocalizedString(@"SHARE_EXTENSION_SENDING_IN_PROGRESS_TITLE", @"Alert title");
+    UIAlertController *progressAlert = [UIAlertController alertControllerWithTitle:progressTitle
+                                                                           message:nil
+                                                                    preferredStyle:UIAlertControllerStyleAlert];
+
+    UIAlertAction *progressCancelAction = [UIAlertAction actionWithTitle:[CommonStrings cancelButton]
+                                                                   style:UIAlertActionStyleCancel
+                                                                 handler:^(UIAlertAction *_Nonnull action) {
+                                                                     [self.shareViewDelegate shareViewWasCancelled];
+                                                                 }];
+    [progressAlert addAction:progressCancelAction];
+
+
+    // Adding a subview to the alert controller like this is a total hack.
+    // ...but it looks good, and given how short a progress view is and how
+    // little the alert controller changes, I'm not super worried about it.
+#ifdef DEBUG
+    if (@available(iOS 12, *)) {
+        // Congratulations! You survived to see another iOS release.
```

Comment:
Can you add the prefix TODO: to this comment

Reply:
Done.